memory/jemalloc_nodump_allocator.cc

```diff
@@ -90,7 +96,30 @@ size_t JemallocNodumpAllocator::UsableSize(void* p,
 
 void* JemallocNodumpAllocator::Allocate(size_t size) {
   int tcache_flag = GetThreadSpecificCache(size);
-  return mallocx(size, MALLOCX_ARENA(arena_index_) | tcache_flag);
+
+  if (arena_indexes_.size() == 1) {
+    return mallocx(size, MALLOCX_ARENA(arena_indexes_[0]) | tcache_flag);
+  }
+
+  // `size` is used as a source of entropy to initialize each thread's arena
+  // selection.
+  //
+  // Core-local may work in place of `thread_local` as we should be able to
+  // tolerate occasional stale reads in thread migration cases. However we need
+  // to switch to std::atomic and prevent cacheline bouncing ourselves. Whether
+  // this is worthwhile is still an open question.
+  thread_local uint32_t tl_rng_seed = static_cast<uint32_t>(size);
+
+  // `tl_rng_seed` is used directly for arena selection and updated afterwards
+  // in order to avoid a close data dependency.
+  void* ret = mallocx(
+      size, MALLOCX_ARENA(
+                arena_indexes_[tl_rng_seed & ((1 << log2_num_arenas_) - 1)]) |
+                tcache_flag);
```

Comment:
nit: you can avoid the repetition and split the code a bit better if you extract the arena selection logic. For example

```
uint32_t JemallocNodumpAllocator::GetArenaIndex(size_t size) const {
  if (arena_indexes_.size() == 1) {
    return 0;
  }
  thread_local ...
   
}

void* JemallocNodumpAllocator::Allocate(size_t size) {
  int tcache_flag = GetThreadSpecificCache(size);
  uint32_t arena_index = GetArenaIndex(size);
  return mallocx(size, MALLOCX_ARENA(arena_index) | tcache_flag);
}
```